services/storage/src/simcore_service_storage/dsm.py

```diff
@@ -426,3 +426,124 @@ def location_from_id(cls, location_id : str):
             destination, filename = _parse_datcore(file_uuid)
             link = await dcw.download_link(destination, filename)
         return link
+
+    async def deep_copy_project_simcore_s3(self, user_id: str, source_project, destination_project, node_mapping):
+        """ Parses a given source project and copies all related files to the destination project
+
+            Since all files are organized as
+
+                project_id/node_id/filename or links to datcore
+
+            this function creates a new folder structure
+
+                project_id/node_id/filename
+
+            and copies all files to the corresponding places.
+
+            Additionally, all external files from datcore are being copied and the paths in the destination
+            project are adapted accordingly
+
+            Lastly, the meta data db is kept in sync
+        """
+        source_folder = source_project["uuid"]
+        dest_folder = destination_project["uuid"]
+
+        # build up naming map based on labels
+        uuid_name_dict = {}
+        uuid_name_dict[dest_folder] = destination_project["name"]
+        for src_node_id, src_node in source_project['workbench'].items():
+            new_node_id = node_mapping.get(src_node_id)
+            if new_node_id is not None:
+                uuid_name_dict[new_node_id] = src_node['label']
+
+        # Step 1: List all objects for this project replace them with the destination object name and do a copy at the same time collect some names
+        _loop = asyncio.get_event_loop()
+        session = aiobotocore.get_session(loop=_loop)
+        async with session.create_client('s3', endpoint_url=self.s3_client.endpoint_url, aws_access_key_id=self.s3_client.access_key,
+            aws_secret_access_key=self.s3_client.secret_key) as client:
+            response = await client.list_objects_v2(Bucket=self.simcore_bucket_name, Prefix=source_folder)
+
+            if "Contents" in response:
+                for f in response['Contents']:
+                    source_object_name = f['Key']
+                    source_object_parts = Path(source_object_name).parts
+
+                    if len(source_object_parts) == 3:
```

Comment:
Error if not content or if not len(source_object_parts) == 3??
At least log int?